client/src/procedures/primitives.rs

```diff
@@ -0,0 +1,1008 @@
+
+#[derive(GuardDebug, Clone, Serialize, Deserialize)]
+pub enum SLIP10DeriveInput {
+    /// Note that BIP39 seeds are allowed to be used as SLIP10 seeds
+    Seed(Location),
+    Key(Location),
+}
+
+/// Derive a SLIP10 child key from a seed or a parent key, store it in output location and
+/// return the corresponding chain code
+#[derive(Procedure, Clone, Serialize, Deserialize)]
+pub struct Slip10Derive {
+    chain: Chain,
+
+    #[output_key]
+    output_key: InterimProduct<OutputKey>,
+
+    #[source]
+    source: SLIP10DeriveInput,
+
+    #[target]
+    target: InterimProduct<Target>,
+}
+
+impl Slip10Derive {
+    pub fn new_from_seed(seed: Location, chain: Chain) -> Self {
+        Self::new(chain, SLIP10DeriveInput::Seed(seed))
+    }
+
+    pub fn new_from_key(parent: Location, chain: Chain) -> Self {
+        Self::new(chain, SLIP10DeriveInput::Key(parent))
+    }
+
+    fn new(chain: Chain, source: SLIP10DeriveInput) -> Self {
+        Slip10Derive {
+            chain,
+            source,
+            target: InterimProduct {
+                target: Target::random(),
+                is_temp: true,
+            },
+            output_key: InterimProduct {
+                target: OutputKey::random(),
+                is_temp: true,
+            },
+        }
+    }
+}
+
+#[execute_procedure]
+impl Process for Slip10Derive {
+    type Input = ();
+    type Output = ChainCode;
+
+    fn process(self, _: Self::Input, guard: GuardedVec<u8>) -> Result<Products<ChainCode>, engine::Error> {
+        let dk = match self.source {
+            SLIP10DeriveInput::Key(_) => {
+                slip10::Key::try_from(&*guard.borrow()).and_then(|parent| parent.derive(&self.chain))
+            }
+            SLIP10DeriveInput::Seed(_) => Seed::from_bytes(&guard.borrow()).derive(Curve::Ed25519, &self.chain),
+        }?;
+        Ok(Products {
+            secret: dk.into(),
+            output: dk.chain_code(),
+        })
+    }
+}
+
+/// Generate a BIP39 seed and its corresponding mnemonic sentence (optionally protected by a
+/// passphrase) and store them in the `output` location
+#[derive(Procedure, Clone, Serialize, Deserialize)]
+pub struct BIP39Generate {
+    passphrase: Option<String>,
+
+    #[target]
+    target: InterimProduct<Target>,
+}
+
+impl BIP39Generate {
+    pub fn new(passphrase: Option<String>) -> Self {
+        BIP39Generate {
+            passphrase,
+            target: InterimProduct {
+                target: Target::random(),
+                is_temp: true,
+            },
+        }
+    }
+}
+
+#[execute_procedure]
+impl Generate for BIP39Generate {
+    type Input = ();
+    type Output = ();
+
+    fn generate(self, _: Self::Input) -> Result<Products<Self::Output>, engine::Error> {
+        let mut entropy = [0u8; 32];
+        fill(&mut entropy)?;
+
+        let mnemonic = bip39::wordlist::encode(
+            &entropy,
+            &bip39::wordlist::ENGLISH, // TODO: make this user configurable
```

Comment:
we should somehow tackle this.

Reply:
I think we should revisit the `BIP39Generate` Procedure in general. At the moment, the generated mnemonic is only used to derive the seed, an then just thrown away. So technically the language doesn't matter, as it is never exposed to the outside, but at the same time this procedure isn't useful for the user, as they have no way to know the mnemonic. wallet.rs for example doesn`t use this procedure, but instead generates the mnemonic outside of stronghold and then uses `Bip39Recover` to create the seed record in stronghold.
I would change the `BIP39Generate` procedure to 1. make the wordlist configurable, and 2. return the mnemonic to the user.

Reply:
I agree. The wordlist can be passed as parameter but as enum, and the resulting mnemonic should be returned to the user.